### DISCUSSION TITLE
API current specification

### DISCUSSION THREAD

**Discussion by vejmarie** [opencomputeproject/CADCloud#19] [API]
The API is not stable, and can evolve pretty fast. Some calls are really not done the way they shall be. Feel free to propose enhancement.

# Projects related calls

### /projects/[username]/getList

This call returns a list of public projects which can be freely accessed when it is not signed and username is not present (optional parameter). If it is signed and present it does return the public and private projects from a specific end user which has signed the call. Signature process follows Amazon s3 version 2 signature process (([https://docs.aws.amazon.com/general/latest/gr/signature-version-2.html](https://docs.aws.amazon.com/general/latest/gr/signature-version-2.html))) 

The signature is pass into the HTTP header of the request. A bash example could be as follow:

dateFormatted=\`TZ=GMT date -R\`
relativePath="/projects/getList"
contentType="application/json"
stringToSign="GET\n\n${contentType}\n${dateFormatted}\n${relativePath}"
signature=\`echo -en ${stringToSign} | openssl sha1 -hmac ${secretKey} -binary | base64\`
curl -s  -X GET \
-H "Host: justyour.parts" \
-H "mydate: ${dateFormatted}" \
-H "Content-Type: ${contentType}" \
-H "Authorization: JYP ${accessKey}:${signature}" \
"https://justyour.parts/projects/getList"

where accessKey and secretKey are provided through the TokenAuth and TokenSecret of the user account.

- Signed: Yes and No
- Input parameters: None
- Return Value: JSON
- Format: Array of Entries, each of them containing: "Name":string, "Owner":string, "Private": Integer, "Date": Array of string [YYYY-MM-DD], "Revisions":Array of Integer 

All following calls could be either unsigned (for public project) or signed for private project.

### /projects/getMagnet/

This call is retrieving the base64 encoded magnet of a specific project which is a png format picture.

- Input parameters: in URL, Date [YYYY-MM-DD], Owner: string, Name: string, Revision: integer
- Example: /projects/getMagnet/2020-01-31/vejmarie/ava/0
 
### /projects/getAvatar/

This call is retrieving the base64 encoded public Avatar from project owner. The return data when decrypted is a png file

- Input parameters: in URL, Date [YYYY-MM-DD], Owner: string, Name: string, Revision: integer
- Example: /projects/getAvatar/2020-01-31/vejmarie/ava/0

### /projects/getPlayerCode/

This call is retrieving the Xeogl code associated with a specific project

- Input parameters: in URL, Date [YYYY-MM-DD], Owner: string, Name: string, Revision: integer
- Example: /projects/getPlayerCode/2020-01-31/vejmarie/ava/0

### /projects/getModel/

This call is retrieving the Wavefront OBJ model of a specific project

- input parameters: in URL, Date [YYYY-MM-DD], Owner: string, Name+"r"+Revision: string, Part: integer

The OBJ file's are split on the backend server due to limitation of Javascript and browser who can handle a maximum size of 64MB strings. If the model is bigger than 16MB it is then split in multi parts.

**Comment by sliptonic** [comment]
With project/getList the behavior seems inconsistent.   If the call is unsigned, I get all projects public on the _server_   But if it is signed, I get all projects for the _user_   

To be consistent the call should be server-centric and should return all projects accessibly _by_ the user.  So an unsigned call would return all projects public on the server and signed call would return all projects public on the server as well as all projects private to the user.

**Reply by vejmarie** [reply]
This would require to modify the API and add a user name as a parameter if I understand properly your requirement. At the present time the unsigned call returns all the project accessible by any end users, which means all public project on the server. When it is signed it does return public and private project from a specific user (the one to whom the credential is assigned to)

**Reply by sliptonic** [reply]
Right but if I have an account on the server, what I want to see is a list of all models that I'm allowed to retrieve from that server.  That means all my own models plus all the public models.  Not just my own private and public models

**Reply by vejmarie** [reply]
Ok, I better understand the need. Let me think about it !

**Reply by sliptonic** [reply]
Are public models always inherently read-only?  The returned list should also include information about whether the model can be written.  I'm imagining a list of models from a server that I have access to.  Some are my own that I can overwrite.  Others are public models that I cannot change.  I should have an icon in the UI that reflects this.

**Reply by vejmarie** [reply]
I looked at the issue during the week-end. As the project list is a moving data (new projects are always created) we have to change a lot of things to return a consistent list. I will start doing that during that from now, as I do have a cleaner vision on how to do it. Public model are always read-only, if a user wants to modify it, a project fork api call will be needed and the project will be copied into user home directory (for modification) or linked (for assembly integration as is). In the case they are links, they will be read-only (maintainer will be the original author, and the original author accept that he can delete the model up to the time no other models are linking to the part)

**Comment by sliptonic** [comment]
/projects/getModel should take a parameter for the desired format.  Default to .obj but also support returning other formats like .gltf or .glb  as well.  .gltf and .glb will require occ 7.5 on the server

**Reply by vejmarie** [reply]
Loved to but at the present time, the backend support only OBJ :). CADCloud is using the snap as FreeCAD to build these files. I updated the snap to OCCT 7.5, but didn't had looked at OBJ export. In fact, CADCloud could virtually export the project into all the format that FreeCAD support. Could you add an enhancement request to the issue tracker as to get somebody looking at the implementation ?

**Comment by sliptonic** [comment]
Does this API discussion include all calls to the server including those initiated by the FreeCAD client?  
I would like the FreeCAD client to be able to retrieve a public model without the user having to create an account on the corresponding server and then populate the token information.

The use case is this:  A user pushes their model to their own CADcloud server, grabs a referencing URL for the model and shares it via chat/email/forum, etc.  A user should be able to copy the link and paste it into a corresponding dialog inside FreeCAD and fetch the model directly into their current document.

This use-case might have implications upstream.  Think about how a google document can be shared;  you can share it with specific people via an email address, or you can share it with 'anyone who has the link'  

The same thing should apply to public projects on a CADcloud server.  I should be able to have a 'public' project which anyone can access with the URL but which don't appear in the public website.

**Reply by vejmarie** [reply]
Yes it would include all calls related from FreeCAD client and the web client. That demand is a top priority on my side and shouldn't be that hard. Question, would you like to retrieve an FCstd file for local storage and opening, or would you like the FreeCAD client just to retrieve the content from the file through the regular Restore call but without any incoming signature as the project would be public.

**Reply by sliptonic** [reply]
I'd like both.  Two separate calls or make it another format of /projects/getModel.

**Reply by vejmarie** [reply]
ok

**Comment by sliptonic** [comment]
Need a call like /projects/getModelInfo/ to get details about a particular model including fetch-able versions
input
-  model ID
- Return value JSON
- Format:  List of:
    - version number
    - date
    - description? (equivalent to git commit message)

**Comment by sliptonic** [comment]
Here's the FreeCAD workbench UI that I'm currently thinking about.
There are three primary use cases:
1) Uploading a model.  The user would make an appropriate selection and use the upload button. A dialog would open for the user to fill out specific information and select a target server.
2) Download by URL.  The user gets a shared url and wants to download it but may not have an account on the server already.  A button opens a dialog where the user can paste the url and the (public) model is fetched into the document
3)  Dock.   This is a tool for browsing the contents of all servers known to the user (either configured through preferences or added by URL)  The user opens the dock and has a combobox to choose from the known servers.  The list below shows models on that server that the user has access to.  

![2021-03-07_20-39](https://user-images.githubusercontent.com/538057/110267375-46cab600-7f85-11eb-922f-6855dd87f9c0.png)

**Reply by vejmarie** [reply]
Seems a good start ;) Let's try to do it !
